### PR TITLE
Playback history

### PR DESCRIPTION
**Pull request by lhc70000**
- Show history in menu
- Show related status in playlist

### CODE REVIEW

**Comment by lhc70000** [comment]
Finished.

File: iina/AppData.swift [hunk 1]
```diff
@@ -90,6 +91,7 @@ struct Constants {
     static let ontopChanged = Notification.Name("IINAOnTopChanged")
     static let keyBindingInputChanged = Notification.Name("IINAkeyBindingInputChanged")
     static let fileLoaded = Notification.Name("IINAFileLoaded")
+    static let historyUpdated = Notification.Name("IINAHistoryUpdated")
   }
   struct Time {
     static let infinite = VideoTime(999, 0, 0)
```
Comment (saagarjha) [inline iina/AppData.swift]: `VideoTime(.max, 0, 0)`

File: iina/HistoryController.swift [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    let result = NSKeyedArchiver.archiveRootObject(history, toFile: plistURL.path)
+    if !result {
+      Utility.log("Cannot save playback history!")
+    }
+    NotificationCenter.default.post(Notification(name: Constants.Noti.historyUpdated))
+  }
+
+  func add(_ url: URL, duration: Double) {
+    guard UserDefaults.standard.bool(forKey: Preference.Key.recordPlaybackHistory) else { return }
+    if let existingItem = (history.filter { $0.mpvMd5 == url.path.md5 }).first, let index = history.index(of: existingItem) {
```
Comment (saagarjha) [inline iina/HistoryController.swift]: `history.first { $0.mpvMd5 == url.path.md5 }`

File: iina/HistoryWindowController.swift [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+      if outlineView.clickedRow >= 0 {
+        indexSet.insert(outlineView.clickedRow)
+      }
+      selectedEntries = indexSet.flatMap { outlineView.item(atRow: $0) as? PlaybackHistory }
+    }
+  }
+
+  override func validateMenuItem(_ menuItem: NSMenuItem) -> Bool {
+    switch menuItem.tag {
+    case 100: // reveal in finder
```
Comment (saagarjha) [inline iina/HistoryWindowController.swift]: I'd create an enum for this.

File: iina/MainWindowMenuActions.swift [hunk 4]
```diff
@@ -421,4 +421,10 @@ extension MainWindowController {
     }
   }
 
+  @IBAction func menuOpenHistory(_ sender: NSMenuItem) {
+    guard let url = sender.representedObject as? URL else { return }
+    playerCore.playFile(url.path)
+    // FIXME: this line shouldn't be here
```
Comment (saagarjha) [inline iina/MainWindowMenuActions.swift]: So, what's the story behind this line?

**Comment by saagarjha** [comment]
@alejx @lhc70000 Looks like I was too late…I'd still recommend going through my comments, though.

File: iina/Extensions.swift [hunk 5]
```diff
@@ -321,6 +321,14 @@ extension Data {
   }
 }
 
+extension String {
```
Comment (lhc70000) [inline iina/Extensions.swift]: Filenames for mpv's watch_later file are calculated by [configfiles.c#L166](https://github.com/mpv-player/mpv/blob/abbc8fc84a99a739e2566e4f7e022b3c1bdde1f3/player/configfiles.c#L166), so I have to implement a md5 method for string in order to get the corresponding watch_later file.

File: iina/MainWindowMenuActions.swift (hunk 4 above)
Comment (lhc70000) [inline iina/MainWindowMenuActions.swift]: `isNetworkResource` should be set in a more general way; not only when opened from history.

File: iina/AppData.swift (hunk 1 above)
Comment (lhc70000) [inline iina/AppData.swift]: IINA may have UI issue when hour goes to 4 digits 😅

I'm planning to move this into `VideoTime`.

File: iina/Extensions.swift (hunk 5 above)
Comment (saagarjha) [inline iina/Extensions.swift]: Alright, but I don't think we need to bridge to `NSString`/`NSData` here. Can't we get this to work with `String`/`Data`?

**Pull request merged**